### PR TITLE
User view model and repository firestore

### PR DESCRIPTION
**Pull request by AlexandreBeeDeeBop**
What is this PR?
This PR introduces a user ViewModel to handle user-related UI logic and a Firestore-backed user repository for managing user data storage and retrieval.

How?
UserViewModel: Manages user data and operations for the UI, providing a clean interface for the view layer.
Firestore UserRepository: Implements data management methods to interact with Firestore, ensuring efficient user data handling.

Why?
To streamline user data operations within the app, enhance the separation of concerns, and leverage Firestore for scalable storage solutions.

Additional Details
Future enhancements will include additional ViewModel functions and Firestore optimizations.

### CODE REVIEW

File: app/src/androidTest/java/com/github/se/cyrcle/model/user/UserTestInstances.kt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    Location(
+        topLeft = Point.fromLngLat(-0.1276, 51.5074), // Top-left corner (London)
+        topRight = Point.fromLngLat(-0.1176, 51.5074), // Top-right corner (London)
+        bottomLeft = Point.fromLngLat(-0.1276, 51.4974), // Bottom-left corner (London)
+        bottomRight = Point.fromLngLat(-0.1176, 51.4974) // Bottom-right corner (London)
+    )
+*/
+
+// Mock Users
+val user1 =
```
Comment (matttsss) [inline app/src/androidTest/java/com/github/se/cyrcle/model/user/UserTestInstances.kt:23]: You could place the `val`s in a `TestInstancesUser` `object`

File: app/src/androidTest/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    userRepositoryFirestore = UserRepositoryFirestore(firestore)
+
+    // Initialize the UserViewModel with the real repository
+    userViewModel = UserViewModel(userRepositoryFirestore)
+    userViewModel.addUser(user1)
+  }
+
+  @Test
+  fun fetchUserById_updates_selectedUser_on_success() = runTest {
+    val countDownLatch = CountDownLatch(1)
```
Comment (matttsss) [inline app/src/androidTest/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt:40]: You can remove the `countDownLatch` it is not used and wastes 5 seconds on the tests

File: app/src/androidTest/java/com/github/se/cyrcle/model/user/UserTestInstances.kt (hunk 1 above)
Comment (LouisJamesVasseur) [reply app/src/androidTest/java/com/github/se/cyrcle/model/user/UserTestInstances.kt:23]: Agreed

File: app/src/androidTest/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    countDownLatch.await(5, TimeUnit.SECONDS)
+
+    // Assert: Check if the user is updated correctly in the StateFlow
+    val fetchedUser = userViewModel.user.first()
+    assertNotNull(fetchedUser)
+    assertEquals("user1", fetchedUser?.userId)
+    assertEquals("john_doe", fetchedUser?.username)
+  }
+
+  val newUser =
```
Comment (matttsss) [inline app/src/androidTest/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt:55]: Those user declaration can be moved into the above mentioned `TestInstancesUser` object

File: app/src/androidTest/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+          assert(user == newUser)
+          countDownLatch.countDown()
+        },
+        { fail("Failed to get User") })
+
+    countDownLatch.await()
+  }
+
+  // Act: Update an existing user
+  val updatedUser =
```
Comment (matttsss) [inline app/src/androidTest/java/com/github/se/cyrcle/model/user/UserViewModelTest.kt:86]: Same as above

File: app/src/main/java/com/github/se/cyrcle/model/user/UserRepositoryFirestore.kt [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+      updatedFavorites.add(parkingId)
+      db.collection(collectionPath)
+          .document(userId)
+          .update("favoriteParkingSpots", updatedFavorites)
+          .addOnSuccessListener { onSuccess() }
+          .addOnFailureListener { onFailure(it) }
+    }
+    .addOnFailureListener { onFailure(it) }
+    */
+    TODO("To fix due to serialization issues")
```
Comment (LouisJamesVasseur) [inline app/src/main/java/com/github/se/cyrcle/model/user/UserRepositoryFirestore.kt:144]: We'll need to fix these serialization issues before integrating the ViewModel into a Screen

**Review by LouisJamesVasseur — APPROVED** [review]
Seems good! We'll need to get back to this class soon to implement all the good ideas you had for additional functions :)

**Comment by matttsss** [comment]
All good! Thanks for applying my sugestions 👌

**Review by matttsss — APPROVED** [review]

**Pull request merged**